src/org/exist/indexing/IndexController.java

```diff
@@ -202,22 +208,33 @@ public void removeCollection(final Collection collection, final DBBroker broker,
     /**
      * Re-index all nodes below the specified root node, using the given mode.
      *
-     * @param transaction the current transaction
+     * @param txn the current transaction
      * @param reindexRoot the node from which reindexing should occur
      * @param mode the mode, one of {@link ReindexMode#UNKNOWN}, {@link ReindexMode#STORE},
      * {@link ReindexMode#REMOVE_SOME_NODES} or {@link ReindexMode#REMOVE_ALL_NODES}.
      */
-    public void reindex(final Txn transaction, IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
+    public void reindex(final Txn txn, final IStoredNode<? extends IStoredNode> reindexRoot, final ReindexMode mode) {
         if (reindexRoot == null) {
             return;
         }
+        IStoredNode<? extends IStoredNode> node = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
+
         setReindexing(true);
-        reindexRoot = broker.objectWith(new NodeProxy(reindexRoot.getOwnerDocument(), reindexRoot.getNodeId()));
-        setDocument(reindexRoot.getOwnerDocument(), mode);
-        getStreamListener();
-        IndexUtils.scanNode(broker, transaction, reindexRoot, listener);
-        flush();
-        setReindexing(false);
+        try {
+            StreamListener stream = getStreamListener(node.getOwnerDocument(), mode);
+            try {
+                stream.startIndexDocument(txn);
+
+                IndexUtils.scanNode(broker, txn, node, listener);
+            } finally {
+                stream.endIndexDocument(txn);
+            }
+
+            //that must be called outside this method to avoid slowdown
```

Comment:
1. `NativeBroker#flush` calls `IndexController#flush` and this still happens from several methods in the index where you haven't removed the calls to `NativeBroker#flush`, i.e. you have only removed some of the calls. 

2. Previously `flush` was called directly from several of the index operations. So now instead of lots of little flushes, we have one big flush when `TransactionManager#processSystemTasks` is eventually called. This has two side-effects:
  1. You cannot RYOW on the index changes until the transaction commits. This could cause data visibility issues.
  2. There is a change in the performance profile, as instead of many small writes we are now doing one big write. I do not know if the change is positive or negative, but I would like to think as you are proposing these changes, you have studied this, and can tell us whether it is positive or negative and what the impact is?